src/client/Microsoft.Identity.Client/ManagedIdentity/AppServiceManagedIdentitySource.cs

```diff
@@ -0,0 +1,98 @@
+        {
+            var msiSecret = EnvironmentVariables.IdentityHeader;
+
+            return TryValidateEnvVars(EnvironmentVariables.IdentityEndpoint, msiSecret, requestContext.Logger, out Uri endpointUri)
+                ? new AppServiceManagedIdentitySource(requestContext, endpointUri, msiSecret)
+                : null;
+        }
+
+        private AppServiceManagedIdentitySource(RequestContext requestContext, Uri endpoint, string secret) : base(requestContext)
+        {
+            _endpoint = endpoint;
+            _secret = secret;
+            _clientId = requestContext.ServiceBundle.Config.UserAssignedClientId;
+            _resourceId = requestContext.ServiceBundle.Config.UserAssignedResourceId;
+        }
+
+        private static bool TryValidateEnvVars(string msiEndpoint, string secret, ILoggerAdapter logger, out Uri endpointUri)
+        {
+            endpointUri = null;
+
+            // if BOTH the env vars endpoint and secret values are null, this MSI provider is unavailable.
+            if (string.IsNullOrEmpty(msiEndpoint) || string.IsNullOrEmpty(secret))
+            {
+                logger.Info("[Managed Identity] App service managed identity is unavailable.");
+                return false;
+            }
+
+            try
+            {
+                endpointUri = new Uri(msiEndpoint);
+            }
+            catch (FormatException ex)
+            {
+                throw new MsalClientException(MsalError.InvalidManagedIdentityEndpoint, string.Format(CultureInfo.InvariantCulture, MsiEndpointInvalidUriError, msiEndpoint), ex);
+            }
+
+            logger.Info($"[Managed Identity] Environment variables validation passed for app service managed identity. Endpoint uri: {endpointUri}");
+            return true;
+        }
+
+        protected override ManagedIdentityRequest CreateRequest(string[] scopes)
+        {
+            // convert the scopes to a resource string
+            string resource = ScopeHelper.ScopesToResource(scopes);
+
+            ManagedIdentityRequest request = new ManagedIdentityRequest(System.Net.Http.HttpMethod.Get, _endpoint);
+            
+            request.Headers.Add(SecretHeaderName, _secret);
+            request.QueryParameters["api-version"] = AppServiceMsiApiVersion;
+            request.QueryParameters["resource"] = resource;
+
+            if (!string.IsNullOrEmpty(_clientId))
+            {
+                _requestContext.Logger.Info("[Managed Identity] Adding user assigned client id to the request.");
+                request.QueryParameters[ClientIdHeaderName] = _clientId;
+            }
+
+            if (!string.IsNullOrEmpty(_resourceId))
+            {
+                _requestContext.Logger.Info("[Managed Identity] Adding user assigned resource id to the request.");
+                request.QueryParameters[Constants.ManagedIdentityResourceId] = _resourceId;
+            }
+
+            request.CreateUri();
```

Comment:
This is a bit strange design. Maybe it's the responsability of `ManagedIdentitySource` to call this instead?